File: vairiot-mobile/app/src/main/java/com/vairiot/app/ui/screens/AuditRunViewModel.kt
package com.vairiot.app.ui.screens

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.vairiot.app.data.api.AuditCampaignResponse
import com.vairiot.app.data.api.AuditReportResponse
import com.vairiot.app.data.api.AuditScanEventResponse
import com.vairiot.app.data.api.RecordScanRequest
import com.vairiot.app.data.api.VairiotApiService
import com.vairiot.app.data.api.ZoneSubmissionResponse
import com.vairiot.app.data.local.QueuedScan
import com.vairiot.app.data.local.QueuedScanDao
import com.vairiot.app.scanner.ScannerService
import com.vairiot.app.sync.ScanSyncScheduler
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

data class AuditRunUiState(
    val campaignId: String = "",
    val campaign: AuditCampaignResponse? = null,
    val isBlind: Boolean = false,
    val isSubmitting: Boolean = false,
    val recentScans:  List<AuditScanEventResponse> = emptyList(),
    val foundCount:   Int = 0,
    val unknownCount: Int = 0,
    val recordedCount: Int = 0,
    val queuedCount:  Int = 0,
    val selectedLocationId: String = "",
    val locations: List<Pair<String, String>> = emptyList(),
    val zoneSubmissions: List<ZoneSubmissionResponse> = emptyList(),
    val condition: String = "",
    val lastMessage:  String? = null,
    val error:        String? = null,
    val report:       AuditReportResponse? = null,
)

@HiltViewModel
class AuditRunViewModel @Inject constructor(
    private val api: VairiotApiService,
    private val scanner: ScannerService,
    private val queuedScanDao: QueuedScanDao,
    private val syncScheduler: ScanSyncScheduler,
    savedStateHandle: SavedStateHandle,
) : ViewModel() {

    private val campaignId: String = savedStateHandle["campaignId"] ?: ""

    private val _state = MutableStateFlow(AuditRunUiState(campaignId = campaignId))
    val state: StateFlow<AuditRunUiState> = _state

    val pendingCount: StateFlow<Int> = queuedScanDao
        .pendingCountByCampaign(campaignId)
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), 0)

    init {
        viewModelScope.launch {
            scanner.scanResults.collect { result -> submitTag(result.value) }
        }
        loadCampaign()
    }

    private fun loadCampaign() {
        viewModelScope.launch {
            val campaign = ensureStarted() ?: return@launch
            val isBlind = campaign.mode == "blind"
            _state.value = _state.value.copy(campaign = campaign, isBlind = isBlind)

            if (isBlind && campaign.siteId != null) {
                try {
                    val locations = api.listSiteLocations(campaign.siteId)
                    val zones = api.listAuditZones(campaignId)
                    _state.value = _state.value.copy(
                        locations = locations.map { it.id to it.name },
                        zoneSubmissions = zones,
                    )
                } catch (_: Exception) {
                    // Zones/locations will be retried on next scan attempt
                }
            }
        }
    }

    // Campaigns are created in 'draft' status; the server only accepts scans and
    // completion once a campaign is 'in_progress'. If it's already in progress
    // (e.g. resuming a partially-scanned audit), starting it conflicts, so fall
    // back to fetching its current state instead.
    private suspend fun ensureStarted(): AuditCampaignResponse? {
        return try {
            api.startAudit(campaignId)
        } catch (_: Exception) {
            try {
                api.listAudits().firstOrNull { it.id == campaignId }
            } catch (_: Exception) {
                null
            }
        }
    }

    fun triggerScan() = scanner.startScan()

    fun setSelectedLocation(locationId: String) {
        _state.value = _state.value.copy(selectedLocationId = locationId)
    }

    fun setCondition(condition: String) {
        _state.value = _state.value.copy(condition = condition)
    }

    fun isZoneLocked(locationId: String): Boolean {
        return _state.value.zoneSubmissions.any { it.locationId == locationId }
    }

    fun submitZone() {
        val locationId = _state.value.selectedLocationId
        if (locationId.isBlank() || campaignId.isBlank()) return
        viewModelScope.launch {
            _state.value = _state.value.copy(isSubmitting = true, error = null)
            try {
                val submission = api.submitAuditZone(campaignId, locationId)
                val updated = _state.value.zoneSubmissions + submission
                _state.value = _state.value.copy(
                    isSubmitting = false,
                    zoneSubmissions = updated,
                    lastMessage = "Zone submitted and locked",
                )
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    isSubmitting = false,
                    error = "Could not submit zone: ${e.message}",
                )
            }
        }
    }

    fun submitTag(tag: String) {
        val trimmed = tag.trim()
        if (trimmed.isBlank() || campaignId.isBlank()) return

        val currentState = _state.value
        if (currentState.isBlind && currentState.selectedLocationId.isBlank()) {
            _state.value = currentState.copy(error = "Select a zone before scanning")
            return
        }
        if (currentState.isBlind && isZoneLocked(currentState.selectedLocationId)) {
            _state.value = currentState.copy(error = "This zone is locked")
            return
        }

        viewModelScope.launch {
            _state.value = _state.value.copy(isSubmitting = true, error = null)
            // Persist the SAME fields the online request sends, so an offline replay
            // isn't rejected for a missing locationId (blind campaigns require it) or
            // silently stripped of its condition assessment.
            val scanLocationId = if (currentState.isBlind) currentState.selectedLocationId else null
            val scanCondition  = currentState.condition.ifBlank { null }
            val queueId = queuedScanDao.insert(
                QueuedScan(
                    campaignId = campaignId,
                    tagValue = trimmed,
                    locationId = scanLocationId,
                    condition = scanCondition,
                ),
            )
            try {
                val request = RecordScanRequest(
                    tagValue = trimmed,
                    locationId = scanLocationId,
                    condition = scanCondition,
                )
                val ev = api.recordAuditScan(campaignId, request)
                queuedScanDao.deleteById(queueId)
                val recents = (listOf(ev) + _state.value.recentScans).take(20)
                val foundDelta    = if (ev.result == "found")    1 else 0
                val unknownDelta  = if (ev.result == "unknown")  1 else 0
                val recordedDelta = if (ev.result == "recorded") 1 else 0
                _state.value = _state.value.copy(
                    isSubmitting  = false,
                    recentScans   = recents,
                    foundCount    = _state.value.foundCount   + foundDelta,
                    unknownCount  = _state.value.unknownCount + unknownDelta,
                    recordedCount = _state.value.recordedCount + recordedDelta,
                    condition     = "",
                    lastMessage   = if (ev.result == "recorded") "Recorded: $trimmed"
                                    else if (ev.result == "found") "Recorded: $trimmed"
                                    else "Unknown tag: $trimmed",
                )
            } catch (e: Exception) {
                syncScheduler.triggerNow()
                _state.value = _state.value.copy(
                    isSubmitting = false,
                    lastMessage  = "Queued offline: $trimmed",
                )
            }
        }
    }

    fun complete() {
        if (campaignId.isBlank()) return
        viewModelScope.launch {
            _state.value = _state.value.copy(isSubmitting = true, error = null)
            try {
                val report = api.completeAudit(campaignId)
                _state.value = _state.value.copy(isSubmitting = false, report = report)
            } catch (e: Exception) {
                _state.value = _state.value.copy(
                    isSubmitting = false,
                    error = "Could not complete audit: ${e.message}",
                )
            }
        }
    }

    fun clearMessage() {
        _state.value = _state.value.copy(lastMessage = null, error = null)
    }
}

File: vairiot-ios/VairiotMobile/Screens/Audits/AuditRunViewModel.swift
import Foundation
import SwiftData

@MainActor
@Observable
final class AuditRunViewModel {

    // MARK: - Scan Result

    enum ScanResult: Equatable {
        case found(assetName: String)
        case unknown(tagValue: String)
        case queued(tagValue: String)
    }

    // MARK: - State

    var audit: AuditCampaignResponse
    var scanCount: Int
    var lastScanResult: ScanResult?
    var zones: [ZoneSubmissionResponse] = []
    var report: AuditReportResponse?
    var isScanning = false
    var isStarting = false
    var isRecording = false
    var isSubmittingZone = false
    var isCompleting = false
    var isLoadingReport = false
    var showScanner = false
    var errorMessage: String?
    var successMessage: String?

    // MARK: - Zone selection

    var selectedZoneLocationId: String?

    // MARK: - Dependencies

    private let apiClient: APIClient

    // MARK: - Init

    init(audit: AuditCampaignResponse, apiClient: APIClient = .shared) {
        self.audit = audit
        self.scanCount = audit.scanCount
        self.apiClient = apiClient
    }

    // MARK: - Start Audit

    func startAudit() async {
        isStarting = true
        errorMessage = nil

        do {
            let updated: AuditCampaignResponse = try await apiClient.request(
                .startAudit(id: audit.id)
            )
            audit = updated
            successMessage = "Audit started"
        } catch let error as APIError {
            errorMessage = error.localizedDescription
        } catch {
            errorMessage = error.localizedDescription
        }

        isStarting = false
    }

    // MARK: - Record Scan

    func recordScan(tagValue: String) async {
        isRecording = true
        errorMessage = nil

        let request = RecordScanRequest(tagValue: tagValue)

        do {
            let event: AuditScanEventResponse = try await apiClient.request(
                .recordAuditScan(campaignId: audit.id, request)
            )
            scanCount += 1

            switch event.result.lowercased() {
            case "found":
                lastScanResult = .found(assetName: event.assetId ?? tagValue)
            default:
                lastScanResult = .unknown(tagValue: tagValue)
            }

            successMessage = "Scan recorded"
        } catch let error as APIError {
            if case .networkError = error {
                // Offline: queue the scan for background sync so it isn't lost.
                queueOfflineScan(tagValue: tagValue)
                scanCount += 1
                lastScanResult = .queued(tagValue: tagValue)
                successMessage = "Offline — scan queued"
            } else {
                errorMessage = error.localizedDescription
            }
        } catch {
            errorMessage = error.localizedDescription
        }

        isRecording = false
    }

    private func queueOfflineScan(tagValue: String) {
        let context = VairiotStore.shared.context
        context.insert(QueuedScan(campaignId: audit.id, tagValue: tagValue))
        try? context.save()
    }

    // MARK: - Submit Zone

    func submitZone(locationId: String) async {
        isSubmittingZone = true
        errorMessage = nil

        do {
            let zone: ZoneSubmissionResponse = try await apiClient.request(
                .submitAuditZone(campaignId: audit.id, locationId: locationId)
            )
            zones.append(zone)
            selectedZoneLocationId = nil
            successMessage = "Zone submitted"
        } catch let error as APIError {
            errorMessage = error.localizedDescription
        } catch {
            errorMessage = error.localizedDescription
        }

        isSubmittingZone = false
    }

    // MARK: - Complete Audit

    func completeAudit() async {
        isCompleting = true
        errorMessage = nil

        do {
            let updated: AuditCampaignResponse = try await apiClient.request(
                .completeAudit(id: audit.id)
            )
            audit = updated
            successMessage = "Audit completed"
            await loadReport()
        } catch let error as APIError {
            errorMessage = error.localizedDescription
        } catch {
            errorMessage = error.localizedDescription
        }

        isCompleting = false
    }

    // MARK: - Load Report

    func loadReport() async {
        isLoadingReport = true

        do {
            report = try await apiClient.request(.getAuditReport(id: audit.id))
        } catch let error as APIError {
            errorMessage = error.localizedDescription
        } catch {
            errorMessage = error.localizedDescription
        }

        isLoadingReport = false
    }

    // MARK: - Load Zones

    func loadZones() async {
        do {
            zones = try await apiClient.request(.listAuditZones(campaignId: audit.id))
        } catch {
            // Non-critical; zones may not exist yet.
        }
    }

    // MARK: - Helpers

    var isActive: Bool {
        audit.status.lowercased() == "in_progress"
    }

    var isDraft: Bool {
        audit.status.lowercased() == "draft"
    }

    var isCompleted: Bool {
        audit.status.lowercased() == "completed"
    }
}
